### PR TITLE
Allow LB to work for apps with IP per task

### LINKED ISSUE #139: Marathon-LB Server Entries don't consider ip-per-container tasks
Since v0.14.0, Marathon has supported launching tasks with their own uniquely addressable  `ipAddress` [\[0\]][0] [\[1\]][1].

Mesos-dns has been updated to check for this information and generate records which take this into account [\[2\]][2]. I'm opening this issue to suggest that Marathon-lb should too.

To expand on what a "relevant record" refers to - the current marathon-lb implementation creates the following mapping for docker bridge tasks:
`<marathon-lb-ip>:<servicePort>` :arrow_right:  `<agent-ip>:<hostPort>` :arrow_right:  `<container-ip>:<containerPort>`

IP-per-container means that the container is routable directly from Marathon-lb. So we can replace the middle entry with the container directly:
`<marathon-lb-ip>:<servicePort>` :arrow_right: `<container-ip>:<discoveryPort>`

(I use docker bridge tasks as an example, but the Mesos task implementation should similarly map directly to the task's ip and port).

The implementation should mirror the one used in the linked mesos-dns PR:
- The presence of the `ipAddress` field should trigger whether this type of entry is created or not. 
- We no longer forward requests to the agent-ip, opting instead for the task's IP which is stored in `app['tasks']['ipAddresses'][0]['ipAddress']`.
    - This field is in a list, but marathon currently only supports 1 ip address per container. So we can always grab the first entry
- We no longer forward to the host port, opting instead to use the container's service port which is specified in `app['discovery']['ports']` (possibly defaulting to 80)

I'm going to close #122 in favor of this issue, as failing design it properly resulted in a fundamentally incorrect implementation. (Marathon restricts tasks which launch with ip-per-container from also specifiying portmappings.)

[0]: https://mesosphere.github.io/marathon/docs/ip-per-task.html
[1]: https://github.com/mesosphere/marathon/releases/tag/v0.14.0
[2]: https://github.com/mesosphere/mesos-dns/pull/197

### PR DESCRIPTION
**Pull request by robbrockbank**
-  Fix up LB to use correct IP address and port.  For IP-per-task, use actual task IP and port.

Fixes #139 

I've added support for IP per task.  There are currently no additional UTs here, but I have fixed up the existing ones.

Note that there is a marathon issue mesosphere/marathon#3636 that will prevent IP-per-tasks from working still, so this is only part of the overall fix required.  Ideally it would be great if someone could get a fix in Marathon for this - I'm a little out of my depth to make that fix.  Could you let me know if is likely to get fixed as it's blocking me.  If not I can provide another PR that works around the missing data in the marathon-lb - but the fix is only useful in certain circumstances.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 .idea
 templates/
 __pycache__
+*.pyc
```
Comment (brndnmtthws) [inline .gitignore]: This is a nit, but small whitespace error here (no newline at EOF).

File: marathon_lb.py [hunk 2]
```diff
@@ -790,6 +815,55 @@ def get_health_check(app, portIndex):
...
+def is_ip_per_task(app):
+    """
+    Return whether the application is using IP-per-task.
+    :param app:  The application to check.
+    :return:  True if using IP per task, False otherwise.
+    """
+    return app.get('ipAddress') is not None
+
+
+def get_task_ip_and_ports(app, task):
```
Comment (brndnmtthws) [inline marathon_lb.py]: Could you write a test around this new function?

**Comment by brndnmtthws** [comment]
Looks great overall. We just need a test around the new function, and we're good to go.

**Comment by robbrockbank** [comment]
@brndnmtthws I've updated the review with some UTs to cover all branches in that method.

What's the take on mesosphere/marathon#3636 ?  Any chance someone could fix that - I'm afraid that one is a little harder for me to get my head around.

**Comment by brndnmtthws** [comment]
I'm not sure about the Marathon issue at this point, but I'll check in with the Marathon team to see if we can get some resources on it.

**Pull request merged**